non-parallel mode broken on master

Looks like https://github.com/shssoichiro/oxipng/commit/4ef92089d5743e896c920d6fa364ff66ab22c346 landed yesterday and removes the non-parallel mode added in #194.

These conditional compilation attributes and a separate synchronous code are important for environments like WebAssembly (which is what they were added for in the first place), where `thread::spawn` is not available and errors out at runtime.

After this change, it's invoked unconditionally again.

I haven't determined what caused this yet, but if I revert 4ef9208, I get these errors:

```
  cargo clippy --no-default-features
    Checking oxipng v2.3.0 (/home/soichiro/repos/oxipng)
warning: unused import: `std::thread`
  --> src/evaluate.rs:23:5
   |
23 | use std::thread;
   |     ^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

error[E0507]: cannot move out of `self.eval_best_candidate.0`, as `self` is a captured variable in an `FnMut` closure
   --> src/evaluate.rs:154:31
    |
104 |     fn try_image_inner(&self, image: Arc<PngImage>, is_reduction: bool) {
    |                        ----- captured outer variable
...
154 |                         match self.eval_best_candidate {
    |                               ^^^^^^^^^^^^^^^^^^^^^^^^ help: consider borrowing here: `&self.eval_best_candidate`
155 |                             Some(prev) if prev.cmp_key() < new.cmp_key() => {}
    |                                  ----
    |                                  |
    |                                  data moved here
    |                                  move occurs because `prev` has type `evaluate::Candidate`, which does not implement the `Copy` trait

error[E0594]: cannot assign to `self.eval_best_candidate` which is behind a `&` reference
   --> src/evaluate.rs:156:34
    |
156 | ...                   _ => self.eval_best_candidate = Some(new),
    |                            ^^^^^^^^^^^^^^^^^^^^^^^^ cannot assign

error: aborting due to 2 previous errors
```

I did attempt to resolve it by leaving the feature-dependent code in place and propagating `&mut` references up the chain, but that did not make the borrow checker very happy.

--- a/src/evaluate.rs
+++ b/src/evaluate.rs
@@ -12,14 +12,15 @@ use crate::png::STD_WINDOW;
 #[cfg(not(feature = "parallel"))]
 use crate::rayon;
 use crate::Deadline;
-#[cfg(feature = "parallel")]
-use rayon;
 use rayon::prelude::*;
+#[cfg(not(feature = "parallel"))]
+use std::cell::RefCell;
 use std::sync::atomic::AtomicUsize;
 use std::sync::atomic::Ordering::SeqCst;
 #[cfg(feature = "parallel")]
 use std::sync::mpsc::*;
 use std::sync::Arc;
+#[cfg(feature = "parallel")]
 use std::thread;
 
 struct Candidate {
@@ -54,7 +55,7 @@ pub(crate) struct Evaluator {
     eval_thread: thread::JoinHandle<Option<Candidate>>,
     // in non-parallel mode, images are evaluated synchronously
     #[cfg(not(feature = "parallel"))]
-    eval_best_candidate: Option<Candidate>,
+    eval_best_candidate: RefCell<Option<Candidate>>,
 }
 
 impl Evaluator {
@@ -70,7 +71,7 @@ impl Evaluator {
             #[cfg(feature = "parallel")]
             eval_thread: thread::spawn(move || rx.into_iter().min_by_key(Candidate::cmp_key)),
             #[cfg(not(feature = "parallel"))]
-            eval_best_candidate: None,
+            eval_best_candidate: RefCell::new(None),
         }
     }
 
@@ -84,7 +85,7 @@ impl Evaluator {
 
     #[cfg(not(feature = "parallel"))]
     fn get_best_candidate(self) -> Option<Candidate> {
-        self.eval_best_candidate
+        self.eval_best_candidate.into_inner()
     }
 
     pub fn get_result(self) -> Option<PngData> {
@@ -151,9 +152,9 @@ impl Evaluator {
 
                     #[cfg(not(feature = "parallel"))]
                     {
-                        match self.eval_best_candidate {
+                        match &mut *self.eval_best_candidate.borrow_mut() {
                             Some(prev) if prev.cmp_key() < new.cmp_key() => {}
-                            _ => self.eval_best_candidate = Some(new),
+                            best => *best = Some(new),
                         }
                     }
                 }

--- a/src/evaluate.rs
+++ b/src/evaluate.rs
@@ -17,6 +17,7 @@ use rayon;
 use rayon::prelude::*;
 use std::sync::atomic::AtomicUsize;
 use std::sync::atomic::Ordering::SeqCst;
+#[cfg(feature = "parallel")]
 use std::sync::mpsc::*;
 use std::sync::Arc;
 use std::thread;
@@ -46,28 +47,44 @@ pub(crate) struct Evaluator {
     nth: AtomicUsize,
     best_candidate_size: Arc<AtomicMin>,
     /// images are sent to the thread for evaluation
+    #[cfg(feature = "parallel")]
     eval_send: SyncSender<Candidate>,
     // the thread helps evaluate images asynchronously
+    #[cfg(feature = "parallel")]
     eval_thread: thread::JoinHandle<Option<Candidate>>,
+    // in non-parallel mode, images are evaluated synchronously
+    #[cfg(not(feature = "parallel"))]
+    eval_best_candidate: Option<Candidate>,
 }
 
 impl Evaluator {
     pub fn new(deadline: Arc<Deadline>) -> Self {
+        #[cfg(feature = "parallel")]
         let (tx, rx) = sync_channel(4);
         Self {
             deadline,
             best_candidate_size: Arc::new(AtomicMin::new(None)),
             nth: AtomicUsize::new(0),
+            #[cfg(feature = "parallel")]
             eval_send: tx,
+            #[cfg(feature = "parallel")]
             eval_thread: thread::spawn(move || rx.into_iter().min_by_key(Candidate::cmp_key)),
+            #[cfg(not(feature = "parallel"))]
+            eval_best_candidate: None,
         }
     }
 
     /// Wait for all evaluations to finish and return smallest reduction
     /// Or `None` if all reductions were worse than baseline.
+    #[cfg(feature = "parallel")]
     fn get_best_candidate(self) -> Option<Candidate> {
         drop(self.eval_send); // disconnect the sender, breaking the loop in the thread
         self.eval_thread.join().expect("eval thread")
+    }
+
+    #[cfg(not(feature = "parallel"))]
+    fn get_best_candidate(self) -> Option<Candidate> {
+        self.eval_best_candidate
     }
 
     pub fn get_result(self) -> Option<PngData> {
@@ -91,6 +108,7 @@ impl Evaluator {
         let best_candidate_size = self.best_candidate_size.clone();
         // sends it off asynchronously for compression,
         // but results will be collected via the message queue
+        #[cfg(feature = "parallel")]
         let eval_send = self.eval_send.clone();
         rayon::spawn(move || {
             let filters_iter = STD_FILTERS.par_iter().with_max_len(1);
@@ -126,7 +144,18 @@ impl Evaluator {
                         nth,
                     };
 
-                    eval_send.send(new).expect("send");
+                    #[cfg(feature = "parallel")]
+                    {
+                        eval_send.send(new).expect("send");
+                    }
+
+                    #[cfg(not(feature = "parallel"))]
+                    {
+                        match self.eval_best_candidate {
+                            Some(prev) if prev.cmp_key() < new.cmp_key() => {}
+                            _ => self.eval_best_candidate = Some(new),
+                        }
+                    }
                 }
             });
         });

